python/ray/experimental/compiled_dag_ref.py

```diff
@@ -148,9 +166,17 @@ def __await__(self):
         fut = self._fut
         self._fut = None
 
-        return_vals = yield from fut.__await__()
+        if self._dag._has_execution_results(self._execution_index):
```

Comment:
Nice change, this looks pretty clean!

One nit: for readability, change the logic to:

```
if not cached:
  yield from fut
  cache values
pop from cache
```